Add data mapping ADR (#27)
OctopusDeploy/Architecture: Decisions/20201014-DataMapping.md

# Overview

> - **Subject**: [explain what decision you are making]

Reviewer: Let's make sure these are filled in before merging.

- **Decision date**: [when was the decision made]

# Executive Summary

> Each explicit data mapping can be defined/implemented variously depending on whether testing, reusability and dependencies are required or not.

Reviewer: This summary does not tell me a lot - what does `variously` mean? We should be summarizing the decision / specifics, and this is a very general statement.

> Each explicit data mapping can be defined/implemented variously depending on whether testing, reusability and dependencies are required or not.

Reviewer: ```suggestion
Data mappings that required dependencies will be defined in classes that can be resolved and injected with those dependencies. Those that do not will be defined in private methods if testing and reuse are not required, or static methods if they are.
```

# Detail

## Considerations

Captured in this document: https://docs.google.com/document/d/1C-4DRUEaER_zWmecM5lECkZAHOie3mlaMWPZFJnfg-k/edit?usp=sharing

## Decision

> If a data mapping requires any dependencies, use a separate class to resolve its dependencies and always use/inject a concrete data mapping type for better discoverability, e.g. [MapToVcsRunbookResources](https://github.com/OctopusDeploy/OctopusDeploy/blob/4702ef0677895d7d41f22af337e0d66f736f787c/source/Octopus.Server/Web/Mapping/MapToVcsRunbookResources.cs#L12). If dependencies are not required, use a private method (e.g. [MapToRunbookDomainModel](https://github.com/OctopusDeploy/OctopusDeploy/compare/spike-tenant-domain-model...spike-tenant-domain-model-runbooks#diff-a5c39abbfe2fc86ceb6451588fa356b4a2d73ed1504d87f1a230686ee6c5bd6eR89)) when both testing (in isolation) and reusability of a data mapping are not necessary, otherwise, use a static method.

Reviewer: ```suggestion
If a data mapping requires any dependencies, use a separate class to resolve its dependencies and always use/inject a concrete data mapping type for better discoverability, e.g. [MapToVcsRunbookResources](https://github.com/OctopusDeploy/OctopusDeploy/blob/4702ef0677895d7d41f22af337e0d66f736f787c/source/Octopus.Server/Web/Mapping/MapToVcsRunbookResources.cs#L12). 

If dependencies are not required, use a private method (e.g. [MapToRunbookDomainModel](https://github.com/OctopusDeploy/OctopusDeploy/compare/spike-tenant-domain-model...spike-tenant-domain-model-runbooks#diff-a5c39abbfe2fc86ceb6451588fa356b4a2d73ed1504d87f1a230686ee6c5bd6eR89)) when testing of the mapping itself, and reuse of the mapping, is not required. Otherwise, use a static method.
```
